tmc-langs-r/src/main/java/fi/helsinki/cs/tmc/langs/r/RPlugin.java

```diff
@@ -0,0 +1,158 @@
+    */
+    private static final Path TEST_FOLDER_PATH = Paths.get("tests");
+    private static final Path TESTTHAT_FOLDER_PATH = Paths.get("testthat");
+    private static final Path TESTTHAT_FILE_PATH = Paths.get("testthat.R");
+
+    private static final String CANNOT_RUN_TESTS_MESSAGE = "Failed to run tests.";
+    private static final String CANNOT_PARSE_TEST_RESULTS_MESSAGE = "Failed to read test results.";
+    private static final String CANNOT_SCAN_EXERCISE_MESSAGE = "Failed to scan exercise.";
+    private static final String CANNOT_PARSE_EXERCISE_DESCRIPTION_MESSAGE =
+            "Failed to parse exercise description.";
+
+    private static Logger log = LoggerFactory.getLogger(RPlugin.class);
+
+    public RPlugin() {
+        super(
+                new ExerciseBuilder(),
+                new StudentFileAwareSubmissionProcessor(),
+                new StudentFileAwareZipper(),
+                new StudentFileAwareUnzipper());
+    }
+
+    /*
+    NOTE: Files.exists does not seem to be able to verify the R and
+    testthat folder's existence if they are empty.
+    */
+    @Override
+    public boolean isExerciseTypeCorrect(Path path) {
+        return Files.exists(path.resolve(R_FOLDER_PATH))
+                || Files.exists(path.resolve(TEST_FOLDER_PATH).resolve(TESTTHAT_FOLDER_PATH))
+                || Files.exists(path.resolve(TEST_FOLDER_PATH).resolve(TESTTHAT_FILE_PATH));
+    }
+
+    @Override
+    protected StudentFilePolicy getStudentFilePolicy(Path projectPath) {
+        return new RStudentFilePolicy(projectPath);
+    }
+
+    @Override
+    public String getPluginName() {
+        return "r";
+    }
+
+    @Override
+    public Optional<ExerciseDesc> scanExercise(Path path, String exerciseName) {
+        ProcessRunner runner = new ProcessRunner(this.getAvailablePointsCommand(), path);
+
+        try {
+            runner.call();
+        } catch (Exception e) {
+            log.error(CANNOT_SCAN_EXERCISE_MESSAGE, e);
+        }
+
+        try {
+            ImmutableList<TestDesc> testDescs = new RExerciseDescParser(path).parse();
+            return Optional.of(new ExerciseDesc(exerciseName, testDescs));
+        } catch (IOException e) {
+            log.error(CANNOT_PARSE_EXERCISE_DESCRIPTION_MESSAGE, e);
+        }
+
+        return Optional.absent();
+    }
+
+    @Override
+    public RunResult runTests(Path path) {
+        ProcessRunner runner = new ProcessRunner(getTestCommand(), path);
+
+        try {
+            runner.call();
+        } catch (Exception e) {
+            log.error(CANNOT_RUN_TESTS_MESSAGE, e);
+        }
+
+        try {
+            return new RTestResultParser(path).parse();
+        } catch (IOException e) {
+            log.error(CANNOT_PARSE_TEST_RESULTS_MESSAGE, e);
+        }
+
+        return null;
+    }
+
+    @Override
+    public ValidationResult checkCodeStyle(Path path, Locale messageLocale) {
+        return null;
```

Comment:
I don't think it's a good idea to return null from this.

Consider returning:
```java
        return new ValidationResult() {
            @Override
            public Strategy getStrategy() {
                return Strategy.DISABLED;
            }

            @Override
            public Map<File, List<ValidationError>> getValidationErrors() {
                return Maps.newHashMap();
            }
        };
```
instead :camel:.

Reply:
I don't know if you've come across this pattern in OhTu. If you are not familiar with these types of `null`-replacing objects, you'll want to Google for "Null Object Pattern".